src/dict.c

```diff
@@ -531,6 +531,55 @@ void *dictFetchValue(dict *d, const void *key) {
     return he ? dictGetVal(he) : NULL;
 }
 
+/* Find an element from the table, also get the plink of the entry. The entry
+ * is returned if the element is found, and the user should later call
+ * `dictFreePlinkEntry` with it in order to unlink and release it. Otherwise if
+ * the key is not found, NULL is returned.
+ *
+ * We can use like this:
+ *
+ * dictEntry *de = dictFindWithPlink(db->dict,key->ptr,&plink, &table);
+ * // Do something, but we can modify the dict
+ * dictFreePlinkEntry(db->dict,de,plink,table); // We don't need to lookup again
+ */
+dictEntry *dictFindWithPlink(dict *d, const void *key, dictEntry ***plink, int *table_index)
+{
+    dictEntry *he, *prevHe;
+    uint64_t h, idx, table;
+
+    if (dictSize(d) == 0) return NULL; /* dict is empty */
+    if (dictIsRehashing(d)) _dictRehashStep(d);
+    h = dictHashKey(d, key);
+
+    for (table = 0; table <= 1; table++) {
+        idx = h & DICTHT_SIZE_MASK(d->ht_size_exp[table]);
+        he = d->ht_table[table][idx];
+        prevHe = NULL;
+        while(he) {
+            if (key==he->key || dictCompareKeys(d, key, he->key)) {
+                if (prevHe) *plink = &prevHe->next;
+                else *plink = &d->ht_table[table][idx];
+                *table_index = table;
+                return he;
```

Comment:
i think we need to increment `d->pauserehash` here, and decrement it on the other method.
and maybe add assertions in dictAdd / dictDelete etc to make sure no one modifies the dict between these two calls.